dot/parachain/types/bitvec.go

```diff
@@ -1,132 +1,239 @@
 	}
 
-	bytes := bv.bytes()
-	_, err = buf.Write(bytes)
+	length := uint(bv.len) // convert to uint for compact encoding
+
+	header, err := scale.Marshal(length)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("marshalling length: %w", err)
 	}
-	return buf.Bytes(), nil
+
+	// Combine the header and the actual bits into the result
+	result := make([]byte, len(header)+len(bv.bits))
+	copy(result, header)                // Copy the header into the result
+	copy(result[len(header):], bv.bits) // Copy the bits into the result
+
+	return result, nil
 }
 
-// UnmarshalSCALE fulfils the SCALE interface for decoding
+// UnmarshalSCALE decodes into the BitVec
 func (bv *BitVec) UnmarshalSCALE(r io.Reader) error {
-	decoder := scale.NewDecoder(r)
-	var size uint
-	err := decoder.Decode(&size)
+	if r == nil {
+		return errors.New("reader is nil")
+	}
+
+	// Read all bytes from reader
+	data, err := io.ReadAll(r)
 	if err != nil {
-		return err
+		return fmt.Errorf("reading data: %w", err)
 	}
-	if size > bitVecMaxLength {
-		return errBitVecTooLong
+
+	// Create a buffer to store the header length
+	var length uint
+	err = scale.Unmarshal(data, &length)
+	if err != nil {
+		return fmt.Errorf("unmarshalling length: %w", err)
 	}
 
-	numBytes := (size + (byteSize - 1)) / byteSize
-	b := make([]byte, numBytes)
-	_, err = r.Read(b)
+	// Check for maximum length
+	if length > MaxBitVecLength {
+		return fmt.Errorf("bitvec length %d exceeds maximum allowed length of %d", length, MaxBitVecLength)
+	}
+
+	// Calculate required bytes for the bits
+	requiredBytes := (int(length) + 7) / 8
+
+	// Get the header size by marshaling the length
+	header, err := scale.Marshal(length)
 	if err != nil {
-		return err
+		return fmt.Errorf("marshalling length: %w", err)
+	}
+	headerSize := len(header)
+
+	// Check if we have enough data
+	if len(data[headerSize:]) < requiredBytes {
+		return fmt.Errorf("incomplete data: got %d bytes, expected %d", len(data[headerSize:]), requiredBytes)
 	}
 
-	bv.setBits(b, size)
+	// Update the BitVec with the decoded data
+	bv.bits = make([]byte, requiredBytes)
+	copy(bv.bits, data[headerSize:headerSize+requiredBytes])
+	bv.len = int(length)
+
 	return nil
 }
+
+// IsEqual checks if two BitVecs are equal by comparing their lengths and bits
+func (bv *BitVec) IsEqual(other *BitVec) bool {
+	// Check if lengths are different
+	if bv.len != other.len {
+		return false
+	}
+
+	// Calculate number of complete bytes to compare
+	completeBytes := bv.len / 8
+
+	// Compare complete bytes first
+	if !bytes.Equal(bv.bits[:completeBytes], other.bits[:completeBytes]) {
+		return false
+	}
+
+	// Check if there are any remaining bits
+	remainingBits := bv.len % 8
+	if remainingBits == 0 {
+		return true
+	}
+
+	// Compare remaining bits in the last byte
+	mask := byte((1 << remainingBits) - 1)
+	lastByteIndex := completeBytes
+	return (bv.bits[lastByteIndex] & mask) == (other.bits[lastByteIndex] & mask)
```

Comment:
given that we ensure that each vectors has the same byte length, is not just compare the raw bytes? any different in the bit position will result in a different byte, so a byte comparision should be enough, no?
```suggestion
        if bv.len != other.len {
		return false
	}
	
	if !bytes.Equal(bv.bits, other.bits) {
		return false
	}
```

Reply:
Yes, also, unused bits will always be zero.
You are right. I will modify it.